File: 437-path-sum-iii/path-sum-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number}
 */
var pathSum = function(root, targetSum) {
    if(root === null) return 0;
    return (
        countPath(root, targetSum) +
        pathSum(root.left, targetSum) +
        pathSum(root.right, targetSum)
    )
};

function countPath(root, targetSum) {
    if(root === null) return 0;

    let count = 0;
    targetSum -= root.val;

    if(targetSum === 0) count++;

    count += countPath(root.left, targetSum);
    count += countPath(root.right, targetSum);

    return count;
}

File: 437-path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if root == None:
            return 0
        return self.countSum(root, targetSum) + self.pathSum(root.left, targetSum) + self.pathSum(root.right, targetSum)

    def countSum(self, root, targetSum) :
        count = 0
        if root == None:
            return 0
        targetSum -= root.val
        if targetSum == 0:
            count+=1
        count+=self.countSum(root.left, targetSum) 
        count+=self.countSum(root.right, targetSum)
        return count
